feat: add bitbucket

beatreichenbach opened the pull request:
This pull request adds support for the Bitbucket remote vcs. Bitbucket doesn't support releases like Github but I added a skeleton class anyway so that users can use `semantic-release version` without having to use the `--no-vcs-release` flag. It still pushes a tag to the remote vcs which creates archives so it's nice to be able to use semantic-release without necessarily having actual releases in the remote vcs.

codejedi365 commented:
This is exciting! Thanks for the work! Recommend adding BitBucket as one of the Hvcs clients tested in:
`tests/unit/semantic_release/changelog.py::test_changelog_context()`
to ensure rendering can occur with the default changelogs.

beatreichenbach commented:
> Recommend adding BitBucket as one of the Hvcs clients tested in: `tests/unit/semantic_release/changelog.py::test_changelog_context()` to ensure rendering can occur with the default changelogs.

Thank you!
I added the test and it all seems to work. (I also have been using it in "production" and the changelog has been created fine.)

beatreichenbach commented:
That's good to know, thanks. I never used different strategies. I'll keep an eye out for that and will play around with different setups to see if i can get anything to break.

codejedi365 reviewed (commented):
Good work on this. A few comments and I'd also recommend adding a few comments to `docs/index.rst` related to the environment variables required for BitBucket.

Otherwise I pass it over to @bernardcooke53 as one of the maintainers.

beatreichenbach commented on tests/unit/semantic_release/hvcs/test_bitbucket.py:
```diff
@@ -0,0 +1,198 @@
+    remote_url,
+    token,
+):
+    with mock.patch.dict(os.environ, patched_os_environ, clear=True):
```
Thank you, I copied everything from the existing tests 🙃

beatreichenbach commented on semantic_release/hvcs/bitbucket.py:
```diff
@@ -0,0 +1,125 @@
+            return self._remote_url
+        if not self.token:
+            raise ValueError("Requested to use token but no token set.")
+        user = os.environ.get("BITBUCKET_USER")
```
Thank you so much for your help with all of this. This is my first real contribution to an open source project, so I really appreciate your detailed explanations, very kind!
I added a section to the `docs/index.rst` to document the only extra environment variable for bitbucket: `BITBUCKET_USER`. Maybe it's not the best solution but it's something until we come up with anything better.

codejedi365 replied:
Congratulations! It's a great feeling and I'm glad you made the endeavor. Glad I could help, and the effort you put forth in adding testing and docs is much more than most do, so keep it up!

bernardcooke53 replied:
I have to echo the sentiment from @codejedi365 - as far as contributions go this one is really excellent @beatreichenbach, particularly for a first-time contribution too!

We will eventually have to do something about the growing number of CI systems and Git remotes. Eventually it would be really nice to rekindle the talk of plugin architectures...

beatreichenbach commented on semantic_release/hvcs/bitbucket.py:
```diff
@@ -0,0 +1,125 @@
+        if hvcs_domain is not None:
+            self.hvcs_domain = hvcs_domain
+        else:
+            api_url = os.getenv("BITBUCKET_SERVER_URL", self.DEFAULT_DOMAIN)
```
You're absolutely right, I copied it from GitHub and removed it, as it makes no sense to use undocumented, unofficial environment variables.

beatreichenbach commented on semantic_release/hvcs/bitbucket.py:
```diff
@@ -0,0 +1,125 @@
+        if hvcs_api_domain is not None:
+            self.hvcs_api_domain = hvcs_api_domain
+        else:
+            api_url = os.getenv("BITBUCKET_API_URL", self.DEFAULT_API_DOMAIN)
```
You're absolutely right, I copied it from GitHub and removed it, as it makes no sense to use undocumented, unofficial environment variables.

bernardcooke53 commented on docs/configuration.rst:
I think the table of default environment variables in `token` needs to be updated too

beatreichenbach replied:
Thanks!
I forgot to run the tests after I removed the environment variables. I removed the environment variable tests and committed the new change. Apologies for missing that!

> I think the table of default environment variables in token needs to be updated too

Can you please point me to that? I'm not sure I understand what you mean.

codejedi365 replied:
It's a documentation change on the configuration page (it's now a table from the merge #786 today). We attempt to guess the `remote.token` value based on the VCS chosen.

beatreichenbach replied:
Thank you again for all the help! This is fantastic.

I rebased my changes (i definitely should have made a branch for this, working in master was not smart.)
And I added the bitbucket token to the table in the docs.

There are a couple lint errors but they should be okay to ignore. Let me know if you want 0 errors and I'll fix those.

bernardcooke53 reviewed (changes requested):
This is a really great piece of work! I'd ask that you take a look at the linter/test failures before merging, but thank you for the effort to make the contribution! 🎉

beatreichenbach commented:
Any updates on this?

beatreichenbach commented:
> LGTM - going to hang fire on merging until the package publishing to PyPI is fixed. Thanks @beatreichenbach!

Thank you so much! No rush on merging.

codejedi365 commented:
@beatreichenbach, the pypi publishing is resolved. Recommend rebasing when you have a chance and updating the lint error that occured.

beatreichenbach commented:
> @beatreichenbach, the pypi publishing is resolved. Recommend rebasing when you have a chance and updating the lint error that occured.

Fixed the ruff linter and rebased onto latest commit.

codejedi365 commented:
@relekang
